instrumentation/github.com/confluentinc/confluent-kafka-go/kafka/splunkkafka/producer.go

```diff
@@ -0,0 +1,148 @@
+
+package splunkkafka
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// A Producer wraps a kafka.Producer and traces its operations.
+type Producer struct {
+	*kafka.Producer
+	cfg            config
+	produceChannel chan *kafka.Message
+}
+
+// NewProducer calls kafka.NewProducer and wraps the resulting Producer with
+// tracing instrumentation.
+func NewProducer(conf *kafka.ConfigMap, opts ...Option) (*Producer, error) {
+	p, err := kafka.NewProducer(conf)
+	if err != nil {
+		return nil, err
+	}
+	return WrapProducer(p, opts...), nil
+}
+
+// WrapProducer wraps a kafka.Producer so that any produced events are traced.
+func WrapProducer(p *kafka.Producer, opts ...Option) *Producer {
+	wrapped := &Producer{
+		Producer: p,
+		cfg:      newConfig(opts...),
+	}
+	wrapped.produceChannel = wrapped.traceProduceChannel(p.ProduceChannel())
+	return wrapped
+}
+
+func (p *Producer) traceProduceChannel(out chan *kafka.Message) chan *kafka.Message {
+	if out == nil {
+		return nil
+	}
+
+	in := make(chan *kafka.Message, cap(out))
+	go func() {
+		for msg := range in {
+			span := p.startSpan(msg)
+			out <- msg
+			span.End()
+		}
+	}()
+
+	return in
+}
+
+func (p *Producer) startSpan(msg *kafka.Message) trace.Span {
+	carrier := NewMessageCarrier(msg)
+	psc := p.cfg.Propagator.Extract(context.Background(), carrier)
+
+	attrs := []attribute.KeyValue{
+		semconv.MessagingSystemKey.String("kafka"),
+		semconv.MessagingDestinationKindTopic,
+		semconv.MessagingDestinationKey.String(*msg.TopicPartition.Topic),
+		semconv.MessagingMessageIDKey.String(strconv.FormatInt(int64(msg.TopicPartition.Offset), 10)), // nolint: gomnd
+		semconv.MessagingKafkaMessageKeyKey.String(string(msg.Key)),
+		semconv.MessagingKafkaPartitionKey.Int64(int64(msg.TopicPartition.Partition)),
+	}
+	opts := []trace.SpanStartOption{
+		trace.WithAttributes(attrs...),
+		trace.WithSpanKind(trace.SpanKindProducer),
+	}
+
+	name := fmt.Sprintf("%s send", *msg.TopicPartition.Topic)
+	ctx, span := p.cfg.Tracer.Start(psc, name, opts...)
+	// Inject the current span into the original message so it can be used to
+	// propagate the span.
+	p.cfg.Propagator.Inject(ctx, carrier)
+	return span
+}
+
+// Close calls the wrapped Producer.Close and closes the producer channel.
+func (p *Producer) Close() {
+	close(p.produceChannel)
+	p.Producer.Close()
+}
+
+// Produce calls the wrapped Producer.Produce and traces the request.
+func (p *Producer) Produce(msg *kafka.Message, deliveryChan chan kafka.Event) error {
+	span := p.startSpan(msg)
+
+	// if the user has selected a delivery channel, we will wrap it and
+	// wait for the delivery event to finish the span
+	if deliveryChan != nil {
+		oldDeliveryChan := deliveryChan
+		deliveryChan = make(chan kafka.Event)
+		go func() {
```

Comment:
should we not have here:

```go
		p.wg.Add(1)
		go func() {
			defer p.wg.Done()
```

?

Reply:
You can create an issue out of it if you do not want to spend time on this right now.